src/lib.rs

```diff
@@ -4,85 +4,108 @@ use pgrx::prelude::*;
 use pgrx::shmem::*;
 use pgrx::spi::SpiResult;
 use pgrx::{debug1, error, pg_shmem_init, GucContext, GucFlags, GucRegistry, GucSetting};
-use serde::{Deserialize, Serialize};
-use std::collections::HashMap;
 use std::ffi::CStr;
+use std::time::Duration;
 
-// Max allocation params
+// Capacity params
+// IMPORTANT: All capacity values MUST be a power of 2. E.g. 2^8 = 256, 2^9 = 512, 2^10 = 1024
 
-const MAX_ENTRIES: usize = 8 * 1024;
-const MAX_CURRENCIES: usize = 1024;
+const MAX_ENTRIES: usize = 512;
+const MAX_CURRENCIES: usize = 64;
+const MAX_ID_PAIRS: usize = 1024;
+const CURRENCY_XUID_MAX_LEN: usize = 64;
 
 // Default Queries
 
 const DEFAULT_Q1_VALIDATION_QUERY: &CStr = cr#"SELECT count(table_name) = 2
 FROM information_schema.tables
 WHERE table_schema = 'plan' AND (table_name = 'currency' or table_name = 'fx_rate');"#;
 
-const DEFAULT_Q2_GET_CURRENCIES_IDS_QUERY: &CStr =
-    cr#"SELECT min(c.id), max(c.id) FROM plan.currency c"#;
-
-const DEFAULT_Q3_GET_CURRENCIES_ENTRY_COUNT: &CStr = cr#"SELECT cu.id,
-LOWER(cu.xuid),
-(SELECT count(*) rate_count FROM plan.fx_rate fr WHERE fr.currency_id = cu.id)
+const DEFAULT_Q2_GET_CURRENCIES_XUID_INIT: &CStr = cr#"SELECT cu.id, LOWER(cu.xuid)
 FROM plan.currency cu
-GROUP by cu.id
 ORDER by cu.id asc;"#;
 
-const DEFAULT_Q4_GET_CURRENCY_ENTRIES: &CStr =
-    cr#"SELECT cr.currency_id, cr.to_currency_id, cr."date", cr.rate
-from plan.fx_rate cr
-order by cr.currency_id asc, cr.to_currency_id asc, cr."date" asc;"#;
+const DEFAULT_Q3_GET_CURRENCY_ENTRIES: &CStr = cr#"WITH
+	dd AS (
+		SELECT date FROM plan.data_date dd
+	),
+	fx_rate_hist AS (
+    	SELECT
+    		cr.currency_id,
+    		cr.to_currency_id,
+    		cr."date",
+    		cr.rate,
+    		ROW_NUMBER() OVER (PARTITION BY currency_id, to_currency_id ORDER BY cr."date" DESC) AS rn
+    	FROM
+    		plan.fx_rate cr
+    		INNER JOIN dd ON cr.date < dd."date"
+	),
+	fx_rate AS (
+    	SELECT
+    		cr.currency_id,
+    		cr.to_currency_id,
+    		cr."date",
+    		cr.rate,
+    		ROW_NUMBER() OVER (PARTITION BY currency_id, to_currency_id ORDER BY cr."date" ASC) AS rn
+    	FROM
+    		plan.fx_rate cr
+    		INNER JOIN dd ON cr.date >= dd."date"
+	)
+SELECT currency_id, to_currency_id, date, rate FROM fx_rate_hist WHERE rn <= 312
+UNION ALL
+SELECT currency_id, to_currency_id, date, rate FROM fx_rate WHERE rn <= 312
+ORDER BY 1, 2, 3;"#;
```

Comment:
lets simplify this query:
WITH
	dd AS (
		SELECT date FROM plan.data_date dd
	),
	fx_rate AS (
    	SELECT
    		cr.currency_id,
    		cr.to_currency_id,
    		cr."date",
    		cr.rate,
    		ROW_NUMBER() OVER (PARTITION BY currency_id, to_currency_id ORDER BY cr."date" DESC) AS rn
    	FROM
    		plan.fx_rate cr
    	ORDER BY
    		cr.currency_id, cr.to_currency_id, cr."date" DESC
	)
SELECT currency_id, to_currency_id, date, rate FROM fx_rate WHERE rn <= 512
ORDER BY 1, 2, 3
;